probe-rs/src/probe/cmsisdap/mod.rs

```diff
@@ -153,6 +221,346 @@ impl CmsisDap {
+                tracing::debug!("JTAG {name} scan chain detected as {n} bits long");
+                n
+            }
+            None => {
+                tracing::error!(
+                    "JTAG {name} scan chain either broken or too long: did not detect 1"
+                );
+                return Err(CmsisDapError::ErrorResponse);
+            }
+        };
+
+        // Check at least one register is detected in the scan chain.
+        if n == 0 {
+            tracing::error!("JTAG {name} scan chain is empty");
+            return Err(CmsisDapError::ErrorResponse);
+        }
+
+        // Check d0[n..] are all 0.
+        if d0[n..].any() {
+            tracing::error!("JTAG {name} scan chain either broken or too long: did not detect 0");
+            return Err(CmsisDapError::ErrorResponse);
+        }
+
+        // Extract d0[..n] as the initial scan chain contents.
+        let data = d0[..n].to_bitvec();
+
+        Ok(data)
+    }
+
+    fn jtag_ensure_shift_dr(&mut self) -> Result<(), CmsisDapError> {
+        // Transition to Test-Logic-Reset.
+        self.jtag_ensure_test_logic_reset()?;
+
+        // Transition to Shift-DR
+        let sequences = vec![
+            JtagSequence::no_capture(false, &bitvec![u8, Lsb0; 0; 1])?,
+            JtagSequence::no_capture(true, &bitvec![u8, Lsb0; 0; 1])?,
+            JtagSequence::no_capture(false, &bitvec![u8, Lsb0; 0; 2])?,
+        ];
+        self.send_jtag_sequences(JtagSequenceRequest::new(sequences)?)?;
+
+        Ok(())
+    }
+
+    fn jtag_ensure_shift_ir(&mut self) -> Result<(), CmsisDapError> {
+        // Transition to Test-Logic-Reset.
+        self.jtag_ensure_test_logic_reset()?;
+
+        // Transition to Shift-IR
+        let sequences = vec![
+            JtagSequence::no_capture(false, &bitvec![u8, Lsb0; 0; 1])?,
+            JtagSequence::no_capture(true, &bitvec![u8, Lsb0; 0; 2])?,
+            JtagSequence::no_capture(false, &bitvec![u8, Lsb0; 0; 2])?,
+        ];
+        self.send_jtag_sequences(JtagSequenceRequest::new(sequences)?)?;
+
+        Ok(())
+    }
+
+    fn send_jtag_configure(&mut self, request: JtagConfigureRequest) -> Result<(), CmsisDapError> {
+        commands::send_command::<JtagConfigureRequest>(&mut self.device, request)
+            .map_err(CmsisDapError::from)
+            .and_then(|v| match v {
+                JtagConfigureResponse(Status::DAPOk) => Ok(()),
+                JtagConfigureResponse(Status::DAPError) => Err(CmsisDapError::ErrorResponse),
+            })
+    }
+
+    fn send_jtag_sequences(
+        &mut self,
+        request: JtagSequenceRequest,
+    ) -> Result<Vec<u8>, CmsisDapError> {
+        commands::send_command::<JtagSequenceRequest>(&mut self.device, request)
+            .map_err(CmsisDapError::from)
+            .and_then(|v| match v {
+                JtagSequenceResponse(Status::DAPOk, tdo) => Ok(tdo),
+                JtagSequenceResponse(Status::DAPError, _) => Err(CmsisDapError::ErrorResponse),
```

Comment:
Can we print additional info here?

Reply:
@Yatekii  What extra info would you like to see here?

Reply:
Just to be clear: would you like to add another error variant (something like `JtagSequenceError`, maybe) that passes along the error code and sequence info up the call stack (to be printed however these errors are already printed), and/or do you suppose we should add a `trace::error` call right here?

One tricky thing either way (at least to my mind; let me know if there's an obvious solution I'm missing) is that a single sequence request can be represented as multiple sub sequences (you need a sequence per change to TMS, and sequences longer than 64 tck cycles also must be split up) -- so I don't know if we just dump the debug format and call it a day, or if we try to construct a more readily grokable textual representation of the sequence, like (spitballing here):

```
Failed to send 5 cycle JTAG sequence; requested logic levels:
    TMS: 11000
    TDI: 01010
```

(where I'm flattening the two sequences implied by the above -- there being at least two sequence because there's the transition from TMS high to TMS low)

I get the motivation to provide more info here in some way, though I'm not sure how to proceed.

Also, one other thing: the Status code is always either `0x00` for success or `0xFF` (excepting invalid response due to buggy probes and/or signal integrity problems), so I don't know if there's much practical utility in surfacing the status when there's an error.

```rust
#[derive(Copy, Clone, Debug)]
pub(crate) enum Status {
    DAPOk = 0x00,
    DAPError = 0xFF,
}

impl Status {
    pub fn from_byte(value: u8) -> Result<Self, SendError> {
        match value {
            0x00 => Ok(Status::DAPOk),
            0xFF => Ok(Status::DAPError),
            _ => Err(SendError::InvalidResponseStatus),
        }
    }
}
```

Reply:
Hmm, you are correct, it's not easily obvious what the correct action is here. I just thought of a simple `trace::debug!() ` call.
Maybe it will never be needed. We can revisit this when we hit an issue. I opened an issue in the meanwhile tracking this. But we can leave it as is. Thanks a lot!